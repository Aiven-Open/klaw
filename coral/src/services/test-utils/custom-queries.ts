import { RenderResult } from "@testing-library/react";
import { within } from "@testing-library/react/pure";

/** Custom query to be able to check for a Definition List
 * that testing-library does not provide
 *
 * `<dl>`: https://www.w3.org/MarkUp/html3/deflists.html
 *
 * @param component RenderResult from testing-library's "render"
 * => `const component = render(<YourComponent)`
 */
function getDefinitionList(component: RenderResult) {
  const list = component.container.querySelectorAll("dl");
  if (list.length === 0) {
    throw Error("No definition list found");
  }

  if (list.length > 1) {
    throw Error("Multiple definition lists found");
  }

  return list[0];
}

/** Custom query to be able to check for a Term in a Definition List
 *
 * `<dt>`: https://www.w3.org/MarkUp/html3/termname.html
 *
 * While testing-library can query for the role "term", this does
 * not confirm that the term is semantically correctly used in
 * a `<dl>` and therefor does not confirm accessibility.
 *
 * @param list a HTMLDListElement, ideally the result from
 * `getDefinitionList`
 */
function getByTermInList(list: HTMLDListElement, term: string) {
  const allTerms = within(list).getAllByRole("term");
  const termComponent = allTerms.filter((termElement) => {
    return termElement.textContent === term;
  });

  if (termComponent === undefined) {
    throw Error(`Could not find a term (<dt>) "${term}"`);
  }

  if (termComponent.length > 1) {
    throw Error(
      `There are multiple term (<dt>) elements with name "${term}". A term should be unique in a definition list!`
    );
  }

  return termComponent[0];
}

function getAllNextDefinitionSiblings(element: HTMLElement) {
  const siblings = [];
  let nextElementSibling = element.nextElementSibling;

  while (nextElementSibling) {
    if (nextElementSibling.tagName === "DD") {
      siblings.push(nextElementSibling);
    }
    nextElementSibling = nextElementSibling.nextElementSibling;
  }
  return siblings;
}

/** Custom query to be able to get all definitions in a Definition List
 *
 * `<dd>`: https://www.w3.org/MarkUp/html3/termdef.html
 *
 * A `<dd>` element is only semantically correct and to identify correctly
 * for assistive technology if it follows a preceding term (`dt`) element.
 * This is why we provide a query to query for all next `dd` siblings of a
 * specific term in a Definition List
 *
 * @param list a HTMLDListElement, ideally the result from
 * `getDefinitionList`
 * @param term the term that the definition relates to
 */
function getAllDefinitions(list: HTMLDListElement, term: string) {
  const termComponent = getByTermInList(list, term);

  const allDDSiblings = getAllNextDefinitionSiblings(termComponent);

  if (allDDSiblings.length === 0) {
    throw Error(`No definitions (<dd>) found related to {term}.`);
  }

  return allDDSiblings;
}

export { getDefinitionList, getByTermInList, getAllDefinitions };
