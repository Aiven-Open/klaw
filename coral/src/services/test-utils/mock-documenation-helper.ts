// createMarkdown and createStringifiedHtml are using unified, which
// provides ES modules that jest can't handle (still). We're mocking
// our helper instead of trying to mock a process etc. to create
// a mock module in __mocks__
// this could be imported in the test file that execute codes that
// is using this functions, but it has to be imported in the right
// order to actually mock the module (before the import that uses
// the function that uses the mocked functions) and this is to brittle
// -> could change through a (automatic) reorder of imports etc.
// so we're importing this in setupAfterEnv to make sure it's mocked
// at the right time
jest.mock("src/domain/helper/documentation-helper", () => ({
  createMarkdown: (string: string | undefined) =>
    `create-markdown-mock-${string}`,
  createStringifiedHtml: (string: string | undefined) =>
    `create-stringified-html-${string}`,
}));
