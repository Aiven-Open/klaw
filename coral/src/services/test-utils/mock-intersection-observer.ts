// This is needed for Design System <Table />
// I followed this approach: https://stackoverflow.com/a/58651649
// we used jest.fn() here before, which lead to errors (that sometimes
// didn't show up in the async tests) because they may get affected
// by e.g. jest.restoreAllMocks etc. in afterX methods
function mockIntersectionObserver({
  root = null,
  rootMargin = "",
  thresholds = [],
  disconnect = () => null,
  observe = () => null,
  takeRecords = () => [],
  unobserve = () => null,
} = {}): void {
  class MockIntersectionObserver implements IntersectionObserver {
    readonly root: Element | null = root;
    readonly rootMargin: string = rootMargin;
    readonly thresholds: ReadonlyArray<number> = thresholds;
    disconnect: () => void = disconnect;
    observe: (target: Element) => void = observe;
    takeRecords: () => IntersectionObserverEntry[] = takeRecords;
    unobserve: (target: Element) => void = unobserve;
  }

  Object.defineProperty(window, "IntersectionObserver", {
    writable: true,
    configurable: true,
    value: MockIntersectionObserver,
  });

  Object.defineProperty(global, "IntersectionObserver", {
    writable: true,
    configurable: true,
    value: MockIntersectionObserver,
  });
}

export { mockIntersectionObserver };
