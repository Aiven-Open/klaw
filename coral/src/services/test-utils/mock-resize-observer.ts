// This is needed for Design System chart components
function mockResizeObserver({
  disconnect = () => null,
  observe = () => null,
  unobserve = () => null,
} = {}): void {
  class MockResizeObserver implements ResizeObserver {
    disconnect: () => void = disconnect;
    observe: (target: Element) => void = observe;
    unobserve: (target: Element) => void = unobserve;
  }

  Object.defineProperty(window, "ResizeObserver", {
    writable: true,
    configurable: true,
    value: MockResizeObserver,
  });

  Object.defineProperty(global, "ResizeObserver", {
    writable: true,
    configurable: true,
    value: MockResizeObserver,
  });
}

export { mockResizeObserver };
