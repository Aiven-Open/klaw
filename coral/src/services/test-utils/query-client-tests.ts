import { QueryClient } from "@tanstack/react-query";

function getQueryClientForTests() {
  return new QueryClient({
    defaultOptions: {
      queries: {
        retry: false,
        cacheTime: Infinity,
      },
    },
  });
}

export { getQueryClientForTests };
