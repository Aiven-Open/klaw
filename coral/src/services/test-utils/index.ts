import { renderWithQueryClient } from "src/services/test-utils/render-with-query-client";

export { renderWithQueryClient };
