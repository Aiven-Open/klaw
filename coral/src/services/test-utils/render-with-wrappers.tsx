import { QueryClientProvider } from "@tanstack/react-query";
import { render, RenderOptions } from "@testing-library/react";
import { ReactElement } from "react";
import { BrowserRouter, MemoryRouter } from "react-router-dom";
import { getQueryClientForTests } from "src/services/test-utils/query-client-tests";

function withQueryClient(ui: ReactElement, options?: RenderOptions) {
  const queryClient = getQueryClientForTests();
  return render(ui, {
    wrapper: ({ children }) => (
      <QueryClientProvider client={queryClient}>{children}</QueryClientProvider>
    ),
    ...options,
  });
}

function withMemoryRouter({
  ui,
  customRoutePath,
  options,
}: {
  ui: ReactElement;
  customRoutePath?: string;
  options?: RenderOptions;
}) {
  return render(ui, {
    wrapper: ({ children }) => (
      <MemoryRouter initialEntries={[customRoutePath || ""]}>
        {children}
      </MemoryRouter>
    ),
    ...options,
  });
}

function withMemoryRouterAndQueryClient({
  ui,
  customRoutePath,
  options,
}: {
  ui: ReactElement;
  customRoutePath?: string;
  options?: RenderOptions;
}) {
  const queryClient = getQueryClientForTests();
  return render(ui, {
    wrapper: ({ children }) => (
      <QueryClientProvider client={queryClient}>
        <MemoryRouter initialEntries={[customRoutePath || ""]}>
          {children}
        </MemoryRouter>
      </QueryClientProvider>
    ),
    ...options,
  });
}

// Only use if BrowserRouter is really needed
// prefer MemoryRouter for tests.
function withBrowserRouterAndQueryClient({
  ui,
  options,
}: {
  ui: ReactElement;
  options?: RenderOptions;
}) {
  const queryClient = getQueryClientForTests();
  return render(ui, {
    wrapper: ({ children }) => (
      <QueryClientProvider client={queryClient}>
        <BrowserRouter>{children}</BrowserRouter>
      </QueryClientProvider>
    ),
    ...options,
  });
}

// To be used for testing URL states (ie, search params)
function withBrowserRouter({
  ui,
  options,
}: {
  ui: ReactElement;
  options?: RenderOptions;
}) {
  return render(ui, {
    wrapper: ({ children }) => <BrowserRouter>{children}</BrowserRouter>,
    ...options,
  });
}

type ValidOptions =
  | "queryClient"
  | "memoryRouter"
  | "browserRouter"
  | "customRoutePath";

type CustomRenderOption = {
  [K in ValidOptions]?: K extends
    | "queryClient"
    | "memoryRouter"
    | "browserRouter"
    ? boolean
    : string;
};

function customRender(
  ui: ReactElement,
  renderWith: CustomRenderOption,
  options?: RenderOptions
) {
  if (renderWith?.queryClient && renderWith?.memoryRouter) {
    return withMemoryRouterAndQueryClient({
      ui,
      customRoutePath: renderWith?.customRoutePath,
      options,
    });
  }

  if (renderWith?.queryClient && renderWith?.browserRouter) {
    return withBrowserRouterAndQueryClient({ ui, options });
  }

  if (renderWith?.memoryRouter) {
    return withMemoryRouter({
      ui,
      customRoutePath: renderWith?.customRoutePath,
      options,
    });
  }

  if (renderWith?.browserRouter) {
    return withBrowserRouter({
      ui,
      options,
    });
  }

  if (renderWith?.queryClient) {
    return withQueryClient(ui, options);
  }

  console.error(
    "You have not passed any renderWith option, which returns the non-custom render."
  );
  return render(ui, options);
}
export { customRender };
