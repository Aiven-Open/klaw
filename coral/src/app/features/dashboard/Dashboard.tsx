import { Alert, Box, Card, Grid, Icon, Template } from "@aivenio/aquarium";
import { AreaChart, Axis, BarChart, Tooltip } from "@aivenio/aquarium/charts";
import loading from "@aivenio/aquarium/icons/loading";
import { useQuery } from "@tanstack/react-query";
import { useEffect, useState } from "react";
import StatsDisplay from "src/app/components/StatsDisplay";
import { useAuthContext } from "src/app/context-provider/AuthProvider";
import { usePendingRequests } from "src/app/hooks/usePendingRequests";
import {
  DashboardsAnalyticsData,
  getActivityLogForTeamOverview,
  getDashboardStats,
} from "src/domain/analytics";
import { parseErrorMsg } from "src/services/mutation-utils";

const Dashboard = () => {
  const {
    data: chartData,
    isLoading: isLoadingChartData,
    isError: isErrorChartData,
    error: errorChartData,
  } = useQuery<DashboardsAnalyticsData, Error>(
    ["getActivityLogForTeamOverview"],
    {
      queryFn: () => getActivityLogForTeamOverview(),
    }
  );

  const {
    data: statsData,
    isLoading: isLoadingStatsData,
    isError: isErrorStatsData,
    error: errorStatsData,
  } = useQuery(["getDashboardStats"], {
    queryFn: getDashboardStats,
  });

  const { TOPIC, ACL, SCHEMA, CONNECTOR } = usePendingRequests();

  const { totalTeamTopics, totalOrgTopics } = useAuthContext();

  // The following contraption is to allow charts to be responsive despite being rendered in a Card
  // Without it, the size of the charts calculated by the ResponsiveContainer they are crapped in under the hood
  // will only be calculated on the first render, which makes the entire Dashboard layout unresponsive on viewport resize
  // This will remove the Chart components and show the loading state while resizing, and then render them again with the proper size
  // Source: https://github.com/recharts/recharts/issues/1423#issuecomment-538670179
  const [forceLoading, setForceLoading] = useState(false);
  useEffect(() => {
    let resizeThrottleTimeout: NodeJS.Timeout;
    let skip = false;
    const resizeCb = () => {
      if (skip) {
        return;
      }

      setForceLoading(true);
      skip = true;
      resizeThrottleTimeout = setTimeout(() => {
        setForceLoading(false);
        skip = false;
      }, 100);
    };
    window.addEventListener("resize", resizeCb);

    return () => {
      clearTimeout(resizeThrottleTimeout);
    };
  }, [setForceLoading]);

  return (
    <Template columns={1} gap={"l2"}>
      <Card title="Topics" fullWidth>
        {isErrorStatsData && (
          <Box marginBottom={"l1"}>
            <Alert type={"error"}>
              Could not gather the topics data: {parseErrorMsg(errorStatsData)}
            </Alert>
          </Box>
        )}
        <Grid cols={"4"}>
          <StatsDisplay
            isLoading={false}
            amount={totalTeamTopics}
            entity={"My team's topics"}
          />
          <StatsDisplay
            isLoading={false}
            amount={totalOrgTopics}
            entity={"My organization's topics"}
          />
          <StatsDisplay
            isLoading={isLoadingStatsData}
            amount={statsData?.producerCount}
            entity={"My producer topics"}
          />
          <StatsDisplay
            isLoading={isLoadingStatsData}
            amount={statsData?.consumerCount}
            entity={"My consumer topics"}
          />
        </Grid>
      </Card>
      <Card title="My team's pending requests" fullWidth>
        <Grid cols={"4"}>
          <StatsDisplay
            isLoading={TOPIC === undefined}
            amount={TOPIC}
            entity={"Topic requests"}
          />
          <StatsDisplay
            isLoading={ACL === undefined}
            amount={ACL}
            entity={"ACL requests"}
          />
          <StatsDisplay
            isLoading={SCHEMA === undefined}
            amount={SCHEMA}
            entity={"Schema requests"}
          />
          <StatsDisplay
            isLoading={CONNECTOR === undefined}
            amount={CONNECTOR}
            entity={"Connector requests"}
          />
        </Grid>
      </Card>

      <Card title="My team's requests per day" fullWidth>
        {isErrorChartData && (
          <Alert type={"error"}>
            Could not gather the requests per day data :{" "}
            {parseErrorMsg(errorChartData)}
          </Alert>
        )}
        {isLoadingChartData || forceLoading ? (
          <Box.Flex
            justifyContent={"center"}
            alignContent={"center"}
            style={{ minHeight: "250px" }}
            flexWrap={"wrap"}
          >
            <Icon icon={loading} fontSize={"30px"} />
          </Box.Flex>
        ) : (
          <AreaChart height={250} data={chartData?.requestsPerDay}>
            <Axis.XAxis dataKey="date" interval={"equidistantPreserveStart"} />
            <Axis.YAxis width={10} />
            <AreaChart.Area dataKey="Requests" />
            <Tooltip />
          </AreaChart>
        )}
      </Card>

      <Card title="My team's topics per environment" fullWidth>
        {isErrorChartData && (
          <Alert type={"error"}>
            Could not gather the topics per environment data:{" "}
            {parseErrorMsg(errorChartData)}
          </Alert>
        )}
        {isLoadingChartData || forceLoading ? (
          <Box.Flex
            justifyContent={"center"}
            alignContent={"center"}
            style={{ minHeight: "250px" }}
            flexWrap={"wrap"}
          >
            <Icon icon={loading} fontSize={"30px"} />
          </Box.Flex>
        ) : (
          <BarChart
            height={250}
            layout={"vertical"}
            data={chartData?.topicsPerEnv}
          >
            <Axis.XAxis dataKey="environment" />
            <Axis.YAxis width={10} />
            {/* fill is --aquarium-colors-primary-100 */}
            <BarChart.Bar dataKey="Topics" fill="#0788d1" />
            <Tooltip />
          </BarChart>
        )}
      </Card>
    </Template>
  );
};

export default Dashboard;
