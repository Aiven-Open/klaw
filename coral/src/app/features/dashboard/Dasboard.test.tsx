import { cleanup, screen } from "@testing-library/react";
import Dashboard from "src/app/features/dashboard/Dashboard";
import { mockResizeObserver } from "src/services/test-utils/mock-resize-observer";
import { customRender } from "src/services/test-utils/render-with-wrappers";

describe("Dashboard.tsx", () => {
  describe("renders correct sections", () => {
    beforeEach(async () => {
      mockResizeObserver();

      customRender(<Dashboard />, {
        memoryRouter: true,
        queryClient: true,
      });
    });

    afterEach(() => {
      cleanup();
    });

    it("renders Topics section", async () => {
      const section = screen.getByText("Topics");

      expect(section).toBeVisible();
    });

    it("renders Pending requests section", async () => {
      const section = screen.getByText("Pending requests");

      expect(section).toBeVisible();
    });

    it("renders Requests per day section", async () => {
      const section = screen.getByText("Requests per day");

      expect(section).toBeVisible();
    });

    it("renders Analytics section", async () => {
      const section = screen.getByText("Analytics");

      expect(section).toBeVisible();
    });
  });
});
