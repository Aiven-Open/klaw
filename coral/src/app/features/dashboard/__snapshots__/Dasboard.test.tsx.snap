// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Dashboard.tsx renders correct DOM according to data received from getRequestsWaitingForApproval, getActivityLogForTeamOverview, getDashboardStats 1`] = `
<div>
  <div
    style="display: grid; gap: 24px; grid-template-columns: 1fr;"
  >
    <div
      class="border-[2px] border-grey-5 rounded-[2px] relative p-5 flex flex-col gap-5 bg-white outline-none transition-all w-full Aquarium-Card"
    >
      <div
        class="flex flex-col gap-4"
      >
        <div
          class="typography-default-strong text-black"
        >
          Topics
        </div>
        <div
          class="typography-caption text-grey-70"
        >
          <div
            style="display: grid; grid-template-columns: repeat(4, minmax(0, 1fr));"
          >
            <div
              style="display: flex; justify-content: space-around; flex-direction: column;"
            >
              <div
                class="typography-heading text-grey-70"
              >
                <span
                  class="visually-hidden"
                >
                  Loading information
                </span>
                <div
                  aria-label="Loading..."
                  class="Aquarium-Skeleton bg-grey-5 h-auto before:content-['_'] whitespace-pre origin-[0%_45%] scale-[0.55] block animate-pulse"
                  role="progressbar"
                />
              </div>
              <div
                class="typography-small text-grey-50"
              >
                My team's topics
              </div>
            </div>
            <div
              style="display: flex; justify-content: space-around; flex-direction: column;"
            >
              <div
                class="typography-heading text-grey-70"
              >
                <span
                  class="visually-hidden"
                >
                  Loading information
                </span>
                <div
                  aria-label="Loading..."
                  class="Aquarium-Skeleton bg-grey-5 h-auto before:content-['_'] whitespace-pre origin-[0%_45%] scale-[0.55] block animate-pulse"
                  role="progressbar"
                />
              </div>
              <div
                class="typography-small text-grey-50"
              >
                My organization's topics
              </div>
            </div>
            <div
              style="display: flex; justify-content: space-around; flex-direction: column;"
            >
              <div
                class="typography-heading text-grey-70"
              >
                <span
                  class="visually-hidden"
                >
                  Loading information
                </span>
                <div
                  aria-label="Loading..."
                  class="Aquarium-Skeleton bg-grey-5 h-auto before:content-['_'] whitespace-pre origin-[0%_45%] scale-[0.55] block animate-pulse"
                  role="progressbar"
                />
              </div>
              <div
                class="typography-small text-grey-50"
              >
                My producer topics
              </div>
            </div>
            <div
              style="display: flex; justify-content: space-around; flex-direction: column;"
            >
              <div
                class="typography-heading text-grey-70"
              >
                <span
                  class="visually-hidden"
                >
                  Loading information
                </span>
                <div
                  aria-label="Loading..."
                  class="Aquarium-Skeleton bg-grey-5 h-auto before:content-['_'] whitespace-pre origin-[0%_45%] scale-[0.55] block animate-pulse"
                  role="progressbar"
                />
              </div>
              <div
                class="typography-small text-grey-50"
              >
                My consumer topics
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
    <div
      class="border-[2px] border-grey-5 rounded-[2px] relative p-5 flex flex-col gap-5 bg-white outline-none transition-all w-full Aquarium-Card"
    >
      <div
        class="flex flex-col gap-4"
      >
        <div
          class="typography-default-strong text-black"
        >
          My team's pending requests
        </div>
        <div
          class="typography-caption text-grey-70"
        >
          <div
            style="display: grid; grid-template-columns: repeat(4, minmax(0, 1fr));"
          >
            <div
              style="display: flex; justify-content: space-around; flex-direction: column;"
            >
              <div
                class="typography-heading text-grey-70"
              >
                <span
                  class="visually-hidden"
                >
                  Loading information
                </span>
                <div
                  aria-label="Loading..."
                  class="Aquarium-Skeleton bg-grey-5 h-auto before:content-['_'] whitespace-pre origin-[0%_45%] scale-[0.55] block animate-pulse"
                  role="progressbar"
                />
              </div>
              <div
                class="typography-small text-grey-50"
              >
                Topic requests
              </div>
            </div>
            <div
              style="display: flex; justify-content: space-around; flex-direction: column;"
            >
              <div
                class="typography-heading text-grey-70"
              >
                <span
                  class="visually-hidden"
                >
                  Loading information
                </span>
                <div
                  aria-label="Loading..."
                  class="Aquarium-Skeleton bg-grey-5 h-auto before:content-['_'] whitespace-pre origin-[0%_45%] scale-[0.55] block animate-pulse"
                  role="progressbar"
                />
              </div>
              <div
                class="typography-small text-grey-50"
              >
                ACL requests
              </div>
            </div>
            <div
              style="display: flex; justify-content: space-around; flex-direction: column;"
            >
              <div
                class="typography-heading text-grey-70"
              >
                <span
                  class="visually-hidden"
                >
                  Loading information
                </span>
                <div
                  aria-label="Loading..."
                  class="Aquarium-Skeleton bg-grey-5 h-auto before:content-['_'] whitespace-pre origin-[0%_45%] scale-[0.55] block animate-pulse"
                  role="progressbar"
                />
              </div>
              <div
                class="typography-small text-grey-50"
              >
                Schema requests
              </div>
            </div>
            <div
              style="display: flex; justify-content: space-around; flex-direction: column;"
            >
              <div
                class="typography-heading text-grey-70"
              >
                <span
                  class="visually-hidden"
                >
                  Loading information
                </span>
                <div
                  aria-label="Loading..."
                  class="Aquarium-Skeleton bg-grey-5 h-auto before:content-['_'] whitespace-pre origin-[0%_45%] scale-[0.55] block animate-pulse"
                  role="progressbar"
                />
              </div>
              <div
                class="typography-small text-grey-50"
              >
                Connector requests
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
    <div
      class="border-[2px] border-grey-5 rounded-[2px] relative p-5 flex flex-col gap-5 bg-white outline-none transition-all w-full Aquarium-Card"
    >
      <div
        class="flex flex-col gap-4"
      >
        <div
          class="typography-default-strong text-black"
        >
          My team's approved requests
        </div>
        <div
          class="typography-caption text-grey-70"
        >
          <div
            class="Aquarium-NativeSelect"
            style="display: flex; flex-direction: column;"
          >
            <label
              class="w-full"
              for="react-aria-:r0:"
              id="react-aria-:r0:-label"
            >
              <span
                class="block mb-2"
              >
                <span
                  class="inline-flex items-center typography-small-strong text-grey-70"
                >
                  Timeframe
                </span>
              </span>
              <span
                class="Aquarium-NativeSelectBase relative block"
              >
                <span
                  class="absolute right-0 inset-y-0 mr-4 text-grey-40 flex ml-3 pointer-events-none typography-default-strong mt-4"
                >
                  <svg
                    aria-hidden="true"
                    class="Aquarium-Icon"
                    data-testid="icon-dropdown"
                    height="1em"
                    preserveAspectRatio="xMidYMid meet"
                    role="img"
                    viewBox="0 0 22 22"
                    width="1em"
                    xmlns="http://www.w3.org/2000/svg"
                    xmlns:xlink="http://www.w3.org/1999/xlink"
                  >
                    <path
                      d="m5.5 8.25 5.5 5.5 5.5-5.5"
                      fill="none"
                      stroke="currentColor"
                      stroke-linecap="round"
                      stroke-linejoin="round"
                      stroke-width="1.5"
                    />
                  </svg>
                </span>
                <select
                  class="block w-full rounded-sm appearance-none disabled:cursor-not-allowed typography-small text-grey-70 placeholder:text-grey-40 focus:outline-none px-3 py-3 disabled:border-grey-20 disabled:bg-grey-5 disabled:text-grey-40 border border-grey-20 hover:border-grey-50 focus:border-info-70"
                  id="react-aria-:r0:"
                >
                  <option
                    value="7"
                  >
                    Last 7 days
                  </option>
                  <option
                    selected=""
                    value="30"
                  >
                    Last 30 days
                  </option>
                  <option
                    value="90"
                  >
                    Last 90 days
                  </option>
                </select>
              </span>
            </label>
            <p
              class="block mt-1 mb-3 typography-caption"
            >
               
            </p>
          </div>
          <div
            style="display: flex; justify-content: center; align-content: center; flex-wrap: wrap; min-height: 250px;"
          >
            <div
              data-icon="<g fill="none" fill-rule="evenodd" stroke="currentColor" stroke-width="3" transform="translate(1 1)"><circle cx="18" cy="18" r="18" stroke-opacity=".5"/><path d="M36 18c0-9.94-8.06-18-18-18"><animateTransform attributeName="transform" dur="1s" from="0 18 18" repeatCount="indefinite" to="360 18 18" type="rotate"/></path></g>"
              data-testid="ds-icon"
            />
          </div>
        </div>
      </div>
    </div>
    <div
      class="border-[2px] border-grey-5 rounded-[2px] relative p-5 flex flex-col gap-5 bg-white outline-none transition-all w-full Aquarium-Card"
    >
      <div
        class="flex flex-col gap-4"
      >
        <div
          class="typography-default-strong text-black"
        >
          My team's topics per environment
        </div>
        <div
          class="typography-caption text-grey-70"
        >
          <div
            style="display: flex; justify-content: center; align-content: center; flex-wrap: wrap; min-height: 250px;"
          >
            <div
              data-icon="<g fill="none" fill-rule="evenodd" stroke="currentColor" stroke-width="3" transform="translate(1 1)"><circle cx="18" cy="18" r="18" stroke-opacity=".5"/><path d="M36 18c0-9.94-8.06-18-18-18"><animateTransform attributeName="transform" dur="1s" from="0 18 18" repeatCount="indefinite" to="360 18 18" type="rotate"/></path></g>"
              data-testid="ds-icon"
            />
          </div>
        </div>
      </div>
    </div>
  </div>
  <div />
</div>
`;
